Compute correct strides after type promotion

nairbv opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28253 Compute correct strides after type promotion**

Summary:

Instead of trying to fix strides after changing dtypes, wait until after
promotion to set them.

fixes: https://github.com/pytorch/pytorch/issues/27824
fixes: https://github.com/pytorch/pytorch/issues/28502

Differential Revision: [D18124950](https://our.internmc.facebook.com/intern/diff/D18124950)

nairbv committed:
Compute correct strides after type promotion

Summary:

Instead of trying to fix strides after changing dtypes, wait until after
promotion to set them.

fixes: https://github.com/pytorch/pytorch/issues/27824

[ghstack-poisoned]

nairbv committed:
Update on "Compute correct strides after type promotion"


Summary:

Instead of trying to fix strides after changing dtypes, wait until after
promotion to set them.

fixes: https://github.com/pytorch/pytorch/issues/27824

[ghstack-poisoned]

nairbv committed:
Update on "Compute correct strides after type promotion"


Summary:

Instead of trying to fix strides after changing dtypes, wait until after
promotion to set them.

fixes: https://github.com/pytorch/pytorch/issues/27824

[ghstack-poisoned]

nairbv committed:
Update on "Compute correct strides after type promotion"


Summary:

Instead of trying to fix strides after changing dtypes, wait until after
promotion to set them.

fixes: https://github.com/pytorch/pytorch/issues/27824

[ghstack-poisoned]

nairbv committed:
Update on "Compute correct strides after type promotion"


Summary:

Instead of trying to fix strides after changing dtypes, wait until after
promotion to set them.

fixes: https://github.com/pytorch/pytorch/issues/27824

[ghstack-poisoned]

nairbv committed:
Update on "Compute correct strides after type promotion"


Summary:

Instead of trying to fix strides after changing dtypes, wait until after
promotion to set them.

fixes: https://github.com/pytorch/pytorch/issues/27824
fixes: https://github.com/pytorch/pytorch/issues/28502

[ghstack-poisoned]

gchanan commented on test/test_type_promotion.py:
```diff
@@ -362,6 +362,18 @@ def f():
+        tmp = torch.ones(9, 9, dtype=torch.float, device=self.device)
+        mask = torch.ones(10, 10, dtype=torch.uint8, device=self.device)
+        result = tmp + mask[1:, 1:]
+        expected = torch.empty(9, 9, dtype=torch.float, device=self.device).fill_(2.)
```
nit: use torch.full with `2.` instead.

nairbv committed:
Update on "Compute correct strides after type promotion"


Summary:

Instead of trying to fix strides after changing dtypes, wait until after
promotion to set them.

fixes: https://github.com/pytorch/pytorch/issues/27824
fixes: https://github.com/pytorch/pytorch/issues/28502

[ghstack-poisoned]
